### ISSUE TITLE
Not Disposing

### ISSUE THREAD

**Issue by buda56** [svishnevsky/PDFtoPrinter#1]
Hi,
    Am using this implementation in a Windows Service and am having issues that the PDFToPrint executable is not being removed from memory. When looking at Task Manager I can see up to 20 instances of the PDFToPrint executable.

Regards.
Peter

**Comment by svishnevsky** [comment]
Hi Peter,

Thanks for your feedback. Could you please tell what package version do you use? 

These 20 instances are running after printing is finished and don't exit at all? The wrapper runs new PDFtoPrinter instance per every Print call. So if you use it in parallel code, it's possible to have a number of processes at the same time but them should exit after document is added to the printing queue. If it isn't, could you please send me a source code example?

Also, based on your feedback, I've release a new version of package with concurrency settings. So you can use wrapper as Singleton to control process instance count.

Thanks,
Sergey

**Comment by buda56** [comment]
Sergey,
                Thanks for the quick response.

The package version I am running is: 1.0.5

To give a bit of background on how I am using your product:

I have written a scheduler application  that runs from a windows service, when the schedule is called it executes any schedules required at that time on a sperate thread. In this instance I am calling a module that checks a database value to see if it needs to do anything, if so then it returns a parameter value that is used to load a Reporting Service Report into memory then uses your package to print this report as a pdf.

I was having issues with the Microsoft Report Viewer and a handles error which I seem to have resolved by forcing Garbage collection (Report Viewer is in a Using statement but didn’t seem to help). Since cleaning up all the PDFtoPrint instances and incorporating the above I have not had another instance as yet. I monitored all yesterday afternoon and have not had a repeat instance, but will continue to monitor today as well.

Here’s the code that creates the report and then calls the print function:
                    using (ReportViewer rvDivider = new ReportViewer())
                    {
                        rvDivider.ProcessingMode = ProcessingMode.Remote;
                        ServerReport serverReport = rvDivider.ServerReport;
                        string myPw = CryptoHelper.Decrypt(rptCredPw);
                        NetworkCredential myCred = new NetworkCredential(rptCredID, myPw, domain);

                        // Get a reference to the report server credentials
                        ReportServerCredentials rsCredentials = serverReport.ReportServerCredentials;

                        // Set the credentials for the server report
                        rsCredentials.NetworkCredentials = myCred;

                        //ReportServerURL is the Web Service URL
                        serverReport.ReportServerUrl = new Uri(reportServer);
                        serverReport.ReportPath = reportPath;

                        //Set Report Parameters
                        ReportParameter BendexID = new ReportParameter();
                        BendexID.Name = "BendexID";
                        BendexID.Values.Add(bID);
                        rvDivider.ServerReport.SetParameters(new ReportParameter[] { BendexID });
                        rvDivider.RefreshReport();

                        // Variables
                        Warning[] warnings;
                        string[] streamIds;
                        string mimeType = string.Empty;
                        string encoding = string.Empty;
                        string extension = string.Empty;

                        byte[] bytes = rvDivider.ServerReport.Render("PDF", null, out mimeType, out encoding, out extension, out streamIds, out warnings);

                        string path = string.Format("{0}\\Bdx_{1}_{2}.pdf", myPath, string.Format("{0:yyMMdd}", DateTime.Now), bxID);
                        FileStream fs = new FileStream(path, FileMode.Create);
                        fs.Write(bytes, 0, bytes.Length);
                        fs.Close();

                        string fileName = path;
                        success = PrintPDF(dividerPrinter, fileName);

                        if (success)
                        {
                            //Set DividerPrinted
                            clsDBHelper.SetDivider(bxID, connString);
                            bxID = "";
                        }
                    }

Here’s the code that prints:
        private bool PrintPDF(string printer, string fileName)
        {
            bool sucessful = false;

            try
            {
                var filePath = fileName;
                var networkPrinter = printer;
                var printTimeOut = new TimeSpan(0, 30, 0);
                var printWrapper = new PDFtoPrintWrapper();
                printWrapper.Print(filePath, networkPrinter, printTimeOut);
                sucessful = true;

            }
            catch (Exception ex)
            {
                sucessful = false;
                string message = string.Format("{0:yyyy/MM/dd HH:mm:ss} RevBend Divider file creation has failed: {1}", DateTime.Now, ex.Message);
                _evLog.WriteToEventLog(message, eventType: EventLogEntryType.Error);
                message = string.Format("{0}{2}{1}", message, ex.StackTrace, Environment.NewLine);
                _errLog.WriteToErrorLog(message, ex.StackTrace, "RevBend Divider file creation");
                _dbLog.LogError(ex, (int)Types.Application.Scheduler, (int)Types.RecordType.Error, Environment.MachineName, "{Scheduler}");
            }

            return sucessful;
        }

Regards
Peter Annandale
IT Support

5-17 Taminga Street
Regency Park, 5010
• 08 8352 0970
•  08 8352 0922
È 0437 647 926
peter.annandale@revroof.com.au<mailto:peter.annandale@revroof.com.au>
www.revolutionroofing.com.au<http://www.revolutionroofing.com.au/>
[cid:image001.jpg@01D2B9F7.73664AB0]
Revolution Roofing Disclaimer
This e-mail and any files transmitted with it are confidential and intended solely for the use of the individual, or entity, to whom they are addressed. If you receive it in error, please let us know by reply e-mail, delete it from your system and destroy the copies. This e-mail is also subject to copyright. No part of it should be reproduced, adapted or transmitted without the written consent of the copyright owner. If you have received this email in error, please delete it from your system and notify the sender immediately. Revolution Roofing does not represent, warrant or guarantee that the integrity of this communication has been maintained, nor that the communication is free of errors, virus or interference. If you have any doubts about the authenticity of an e-mail purportedly sent by us, please contact us immediately.

**Comment by buda56** [comment]
Sergey,
       We have found another issue as well, it would appear that when mulitiple print jobs are called not all jobs are printing, so if the loop count is > 1 then the print issues  appear.

Regards
Peter

**Comment by svishnevsky** [comment]
Hi Peter,

Thanks for detailed information, this really helped.
the "Print" is asynchronous and you uses it in synchronous code. Could you try to change the method's call like this:
`printWrapper.Print(filePath, networkPrinter, printTimeOut).Wait();`
to be ensure that printing has started?

**Comment by buda56** [comment]
Sergey,
                Thanks, I’ll implement and let you know the results next week.

Regards
Peter Annandale
IT Support

5-17 Taminga Street
Regency Park, 5010
• 08 8352 0970
•  08 8352 0922
È 0437 647 926
peter.annandale@revroof.com.au<mailto:peter.annandale@revroof.com.au>
www.revolutionroofing.com.au<http://www.revolutionroofing.com.au/>
[cid:image001.jpg@01D2B9F7.73664AB0]
Revolution Roofing Disclaimer
This e-mail and any files transmitted with it are confidential and intended solely for the use of the individual, or entity, to whom they are addressed. If you receive it in error, please let us know by reply e-mail, delete it from your system and destroy the copies. This e-mail is also subject to copyright. No part of it should be reproduced, adapted or transmitted without the written consent of the copyright owner. If you have received this email in error, please delete it from your system and notify the sender immediately. Revolution Roofing does not represent, warrant or guarantee that the integrity of this communication has been maintained, nor that the communication is free of errors, virus or interference. If you have any doubts about the authenticity of an e-mail purportedly sent by us, please contact us immediately.

**Comment by buda56** [comment]
Hi,
                Have implemented your latest release and also included the wait()

Seems to now be working as expected, thanks

Regards
Peter Annandale
IT Support

5-17 Taminga Street
Regency Park, 5010
• 08 8352 0970
•  08 8352 0922
È 0437 647 926
peter.annandale@revroof.com.au<mailto:peter.annandale@revroof.com.au>
www.revolutionroofing.com.au<http://www.revolutionroofing.com.au/>
[cid:image001.jpg@01D2B9F7.73664AB0]
Revolution Roofing Disclaimer
This e-mail and any files transmitted with it are confidential and intended solely for the use of the individual, or entity, to whom they are addressed. If you receive it in error, please let us know by reply e-mail, delete it from your system and destroy the copies. This e-mail is also subject to copyright. No part of it should be reproduced, adapted or transmitted without the written consent of the copyright owner. If you have received this email in error, please delete it from your system and notify the sender immediately. Revolution Roofing does not represent, warrant or guarantee that the integrity of this communication has been maintained, nor that the communication is free of errors, virus or interference. If you have any doubts about the authenticity of an e-mail purportedly sent by us, please contact us immediately.

**Comment by svishnevsky** [comment]
Hi,

That great! Feel free to contact me with any issues.
